jib-plugins-common/src/main/java/com/google/cloud/tools/jib/plugins/common/SingleThreadedLogger.java

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright 2018 Google LLC.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.cloud.tools.jib.plugins.common;
+
+import com.google.common.util.concurrent.Futures;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.function.Consumer;
+
+/**
+ * Keeps all log messages in a sequential, deterministic order along with an additional footer that
+ * always appears below log messages.
+ */
+class SingleThreadedLogger {
+
+  /** ANSI escape sequence for moving the cursor up one line. */
+  private static final String CURSOR_UP_SEQUENCE = "\033[1A";
+
+  private final ExecutorService executorService = Executors.newSingleThreadExecutor();
+
+  private final Consumer<String> plainLogger;
+
+  private String footer = "";
+  private int footerLineCount = 0;
+
+  SingleThreadedLogger(Consumer<String> plainLogger) {
+    this.plainLogger = plainLogger;
+  }
+
+  /**
+   * Runs {@code messageLogger} asynchronously.
+   *
+   * @param messageLogger the {@link Runnable} intended to synchronously log a message to the
+   *     console
+   * @return a {@link Future} to track completion
+   */
+  public Future<Void> log(Runnable messageLogger) {
+    return log(messageLogger, footer, footerLineCount);
+  }
+
+  /**
+   * Sets the footer asynchronously.
```

Comment:
It can also output to the `plainLogger` (whatever that means, erasing the previous output?)